File: src/main/kotlin/io/titandata/titan/providers/local/RemoteLog.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.local

import io.titandata.client.apis.RemotesApi
import io.titandata.client.infrastructure.ClientException
import io.titandata.serialization.RemoteUtil
import io.titandata.remote.nop.NopRemote
import io.titandata.models.Remote

class RemoteLog (
        private val exit: (message: String, code: Int) -> Unit,
        private val remotesApi: RemotesApi = RemotesApi(),
        private val remoteUtil: RemoteUtil = RemoteUtil()
) {
    private val n = System.lineSeparator()
    
    private fun getRemotes(container: String, remoteName: String?): Array<Remote> {
        return try {
            when (remoteName.isNullOrBlank()) {
                true -> remotesApi.listRemotes(container)
                else -> arrayOf(remotesApi.getRemote(container, remoteName))
            }
        } catch (e: ClientException) {
            arrayOf(NopRemote( name="NOP"))
        }
    }

    fun remoteLog(container:String, remoteName: String?, tags: List<String>) {
        val remotes = getRemotes(container, remoteName)
        if(remotes.isEmpty()) {
            exit("remote is not set, run 'remote add' first", 1)
        }

        var first = true
        loop@ for (remote in remotes) {
            if (remote.provider == "nop") {
                break@loop
            }
            try {
                if (!first) {
                    println("")
                } else {
                    first = false
                }
                val commits = remotesApi.listRemoteCommits(container, remote.name, remoteUtil.getParameters(remote), tags)
                for (commit in commits) {
                    println("Remote: ${remote.name}")
                    println("Commit ${commit.id}")
                    if (commit.properties.containsKey("author")) {
                        println("Author: ${commit.properties["author"]}")
                    }
                    if (commit.properties.containsKey("user")) {
                        println("User: ${commit.properties["user"]}")
                    }
                    if (commit.properties.containsKey("email")) {
                        println("Email: ${commit.properties["email"]}")
                    }
                    println("Date:   ${commit.properties["timestamp"]}")
                    if (commit.properties.containsKey("tags")) {
                        val tags = commit.properties.get("tags") as Map<String, String>
                        if (!tags.isEmpty()) {
                            print("Tags:")
                            for ((key, value) in tags) {
                                print(" ")
                                if (value != "") {
                                    print("$key=$value")
                                } else {
                                    print(key)
                                }
                            }
                            println("")
                        }
                    }
                    if (commit.properties["message"] != "") {
                        println("${n}${commit.properties["message"]}")
                    }
                }
            } catch (e: ClientException) {
                println("${remote.name} has not been initialized.")
            }
        }
    }
}

File: internal/app/providers/common/RemoteLog.go
package common

import (
	"fmt"
	"github.com/antihax/optional"
	"github.com/titan-data/remote-sdk-go/remote"
	client "github.com/titan-data/titan-client-go"
	"os"
	"strconv"
)

func getRemotes(repo string, remoteName string) []client.Remote {
	var remotes []client.Remote
	if remoteName != "" {
		r, _, _ := remotesApi.GetRemote(ctx, repo, remoteName)
		remotes = append(remotes, r)
	} else {
		remotes, _, _ = remotesApi.ListRemotes(ctx, repo)
	}
	return remotes
}

func RemoteLog(repo string, remoteName string, tags []string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	remotes := getRemotes(repo, remoteName)
	if len(remotes) == 0 {
		fmt.Println("remote is not set, run 'remote add' first")
		os.Exit(1)
	}
	first := true
	for _, r := range remotes {
		gp, _ := remote.Get(r.Provider).GetParameters(r.Properties)
		p := client.RemoteParameters{
			Provider:   r.Provider,
			Properties: gp,
		}
		o := optional.NewInterface(tags)
		opts := client.ListRemoteCommitsOpts{Tag:o}
		commits, _, err := remotesApi.ListRemoteCommits(ctx, repo, r.Name, p, &opts)
		if err == nil {
			for _, c := range commits {
				if !first {
					fmt.Println()
				} else {
					first = false
				}
				fmt.Println("Commit " + c.Id)
				ifContainsPrint(c.Properties, "author")
				ifContainsPrint(c.Properties, "user")
				ifContainsPrint(c.Properties, "email")
				ifContainsPrint(c.Properties, "timestamp")
				remoteTags, ok := c.Properties["tags"].(map[string]interface{})
				if ok {
					fmt.Print("Tags: ")
					for t, v := range remoteTags {
						if len(v.(string)) > 0 {
							fmt.Printf("%v=%v ", t, v)
						} else  {
							fmt.Printf("%v ", t)
						}
					}
					fmt.Println()
				}
				ifContainsPrint(c.Properties, "message")
			}
		} else {
			fmt.Println( r.Name + " has not been initialized.")
		}
	}
}
